### ISSUE TITLE
Missing xmlresolver-5.1.1.jar causes failure in parsing xpath suppressions

### ISSUE THREAD

**Issue by Fernal73** [checkstyle/checkstyle#12969]
Checkstyle version: 10.9.3

Parsing of xpath suppressions file fails when running checkstyle ant task because of missing 
xmlresolver-5.1.1.jar in the transitive dependencies for saxon-HE.
---

**Describe what you expect in detail.**
The above jar file must be present in the classpath generated.

Using ivy to generate the classpath.

https://github.com/Fernal73/LearnJava/tree/main/Netty

---

**Comment by romani** [comment]
Share exact xpath string that causes a problem

**Comment by Fernal73** [comment]
This is the error thrown by ant under Netty:

```
BUILD FAILED
/root/LearnJava/Netty/build.xml:369: java.lang.NoClassDefFoundError: org/xmlresolver/Resolver
        at net.sf.saxon.lib.CatalogResourceResolver.<init>(CatalogResourceResolver.java:46)             at net.sf.saxon.Configuration.init(Configuration.java:386)
        at net.sf.saxon.Configuration.<init>(Configuration.java:233)                                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)                          at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)          at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)                at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:306)                                                     at java.base/java.lang.Class.newInstance(Class.java:684)                                        at net.sf.saxon.Configuration.newConfiguration(Configuration.java:249)                          at com.puppycrawl.tools.checkstyle.filters.XpathFilterElement.<init>(XpathFilterElement.java:139)                                               at com.puppycrawl.tools.checkstyle.filters.XpathFilterElement.<init>(XpathFilterElement.java:89)                                                at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getXpathFilter(SuppressionsLoader.java:202)                                       at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.startElement(SuppressionsLoader.java:146)                                         at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:518)                               at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractXMLDocumentParser.emptyElement(AbstractXMLDocumentParser.java:183)
        at java.xml/com.sun.org.apache.xerces.internal.impl.dtd.XMLDTDValidator.emptyElement(XMLDTDValidator.java:752)                                  at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanStartElement(XMLDocumentFragmentScannerImpl.java:1387)                                                   at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)                                         at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)                                                      at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)                                     at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)                                   at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:81)                at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)                                 at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)                                at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)                                at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:583)                             at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:203)                                              at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:123)                   at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)                                              at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:475)                         at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)                                              at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:402)                                           at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.realExecute(CheckstyleAntTask.java:298)                                                at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.execute(CheckstyleAntTask.java:280)                                                    at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:299)                         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)                                        at java.base/java.lang.reflect.Method.invoke(Method.java:578)                                   at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)                   at org.apache.tools.ant.Task.perform(Task.java:350)                                             at org.apache.tools.ant.Target.execute(Target.java:449)                                         at org.apache.tools.ant.Target.performTasks(Target.java:470)                                    at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1401)                         at org.apache.tools.ant.Project.executeTarget(Project.java:1374)                                at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)          at org.apache.tools.ant.Project.executeTargets(Project.java:1264)                               at org.apache.tools.ant.Main.runBuild(Main.java:818)                                            at org.apache.tools.ant.Main.startAnt(Main.java:223)                                            at org.apache.tools.ant.launch.Launcher.run(Launcher.java:284)
        at org.apache.tools.ant.launch.Launcher.main(Launcher.java:101)
Caused by: java.lang.ClassNotFoundException: org.xmlresolver.Resolver
        at org.apache.tools.ant.AntClassLoader.findClassInComponents(AntClassLoader.java:1402)          at org.apache.tools.ant.AntClassLoader.findClass(AntClassLoader.java:1357)                      at org.apache.tools.ant.AntClassLoader.loadClass(AntClassLoader.java:1112)                      at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 53 more
```

The classpath is generated by LearnJava/build.xml in file cs.cp generated by the build.

I've added a workaround to append xmlresolver.cp to the above to resolve this for now.

**Comment by romani** [comment]
Please reproduce this problem by our CLI

**Comment by Fernal73** [comment]
> Please reproduce this problem by our CLI

With your checkstyleall.jar? Where do I locate it?

**Comment by Fernal73** [comment]
Tried it with :
```bash
java -jar checkstyle-10.9.3-all.jar -c=../csrules.xml -p=../checkstyle.properties netty/netty/time/UnixTime.java

```
No problem with that.

It seems to be coming from ivy while generating transitive dependencies for Saxon-HE.

**Comment by rnveach** [comment]
@romani We have no XML examples at https://github.com/sevntu-checkstyle/checkstyle-samples to show our side works with ant/ivy/gradle.

**Comment by romani** [comment]
Ivy should be configured to download all dependencies.

**Comment by Fernal73** [comment]
I must be missing something. As far as I can see, your example ivy also doesn't download the xmlresolver-5.1.1.jar.

**Comment by romani** [comment]
Extend our example project to use xpath that you have problem with to show us a problem.

**Comment by Fernal73** [comment]
> Ivy should be configured to download all dependencies and use them during runtime. Jar in maven does no have all dependencies in it. User jar , with all dependencies, is at our GitHub release page.

So maven is not reliable to obtain all required dependencies. The same configuration used to work earlier. Something has changed with Saxon-HE. It's not downloading the xmlresolver.jar , just the xmlresolver-data.jar

**Comment by Fernal73** [comment]
> Extend our example project to use xpath that you have problem with to show us a problem.

I could do that ... in some time. It would be faster for you to simply generate xpath suppressions for your example and run checkstyle on it again. You will need to configure the config file to support xpath suppressions.

**Comment by Fernal73** [comment]
> Ivy should be configured to download all dependencies and use them during runtime. Jar in maven does no have all dependencies in it. User jar , with all dependencies, is at our GitHub release page.

How do you construct your uber jar?

**Comment by Fernal73** [comment]
> > Ivy should be configured to download all dependencies and use them during runtime. Jar in maven does no have all dependencies in it. User jar , with all dependencies, is at our GitHub release page.
> 
> So maven is not reliable to obtain all required dependencies. The same configuration used to work earlier. Something has changed with Saxon-HE. It's not downloading the xmlresolver.jar , just the xmlresolver-data.jar

From what I can glean so far, Ivy seems to skip the main jar while resolving the transitive dependencies for saxon-HE and retrieves only the classified data jar. 

Classified jars have to be explicitly requested using ivy.  Not so for the default jar.

https://stackoverflow.com/questions/24943065/ivyinstall-from-maven-with-classifiers

**Comment by Fernal73** [comment]
Checkstyle doesn't handle the case when the suppressions file is empty.

**Comment by romani** [comment]
If you can reproduce empty file problem by CLI, please create new issue

**Comment by Fernal73** [comment]
> If you can reproduce empty file problem by CLI, please create new issue .
> 
> It should be still valid xml header.

Run gen-suppressions in succession twice on the project w/o deleting the output from first run. The second run creates an empty suppressions file. Run checkstyle task and the process throws an exception as follows:

```
/root/CheckStyleSample/build.xml:15: Unable to create Root Module: config {checkstyle.xml}.             at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:405)                                           at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.realExecute(CheckstyleAntTask.java:298)                                                at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.execute(CheckstyleAntTask.java:280)                                                    at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:299)                         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)                                        at java.base/java.lang.reflect.Method.invoke(Method.java:578)                                   at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)                   at org.apache.tools.ant.Task.perform(Task.java:350)                                             at org.apache.tools.ant.Target.execute(Target.java:449)                                         at org.apache.tools.ant.Target.performTasks(Target.java:470)                                    at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1401)                         at org.apache.tools.ant.Project.executeTarget(Project.java:1374)                                at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)          at org.apache.tools.ant.Project.executeTargets(Project.java:1264)                               at org.apache.tools.ant.Main.runBuild(Main.java:818)                                            at org.apache.tools.ant.Main.startAnt(Main.java:223)                                            at org.apache.tools.ant.launch.Launcher.run(Launcher.java:284)                                  at org.apache.tools.ant.launch.Launcher.main(Launcher.java:101)                         Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.                             at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:480)                         at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)                                              at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:402)                                           ... 17 more                             Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.                   at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:128)                   at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:475)                         ... 19 more                             Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:292)                                 at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)                                at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:583)                             at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:203)                                              at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:123)                   ... 21 more                             Caused by: org.xml.sax.SAXParseException; systemId: file:/root/CheckStyleSample/suppressions-xpath.xml; lineNumber: 1; columnNumber: 1; Premature end of file.                                          at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)                                at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)                                     at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)                                     at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1465)                                           at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:1013)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)                                at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)                                                      at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)                                     at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)                                   at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:81)                at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)                                 ... 26 more
```

**Comment by rnveach** [comment]
> Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse suppressions-xpath.xml - Premature end of file.

This doesn't look like an ant/ivy issue.

> Run gen-suppressions in succession twice 

Does the suppression generation work if you run it only once? Can you provide the contents of the file when the error occurs?

**Comment by rnveach** [comment]
> You may try https://github.com/Fernal73/CheckStyleSample/ to reproduce the error.

I cloned the repo and got no error on running `ant` and it downloaded the xml resolver jar.

````
PS Z:\> git clone https://github.com/Fernal73/CheckStyleSample.git
Cloning into 'CheckStyleSample'...
remote: Enumerating objects: 118, done.
remote: Counting objects: 100% (118/118), done.
remote: Compressing objects: 100% (91/91), done.
remote: Total 118 (delta 17), reused 112 (delta 11), pack-reused 0Receiving objects:  85% (101/118), 37.29 MiB | 9
Receiving objects: 100% (118/118), 40.75 MiB | 9.43 MiB/s, done.
Resolving deltas: 100% (17/17), done.
PS Z:\> cd .\CheckStyleSample\
PS Z:\CheckStyleSample> ant
...
[ivy:retrieve]  found org.xmlresolver#xmlresolver;5.1.1 in local-maven-2
...
[ivy:retrieve]  59 artifacts copied, 0 already retrieved (43465kB/1715ms)

init-checkstyle:

checkstyle:
[checkstyle] Running Checkstyle 10.9.3 on 1 files
[checkstyle] Running Checkstyle 10.9.3 on 1 files

makepom:

main:

BUILD SUCCESSFUL
Total time: 8 seconds
````

**Comment by Fernal73** [comment]
> > If you can reproduce empty file problem by CLI, please create new issue .
> > It should be still valid xml header.
> 
> Run gen-suppressions in succession twice on the project w/o deleting the output from first run. The second run creates an empty suppressions file. Execute checkstyle task and the process throws an exception as follows:
> 
> ```
> /root/CheckStyleSample/build.xml:15: Unable to create Root Module: config {checkstyle.xml}.             at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:405)                                           at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.realExecute(CheckstyleAntTask.java:298)                                                at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.execute(CheckstyleAntTask.java:280)                                                    at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:299)                         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)                                        at java.base/java.lang.reflect.Method.invoke(Method.java:578)                                   at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)                   at org.apache.tools.ant.Task.perform(Task.java:350)                                             at org.apache.tools.ant.Target.execute(Target.java:449)                                         at org.apache.tools.ant.Target.performTasks(Target.java:470)                                    at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1401)                         at org.apache.tools.ant.Project.executeTarget(Project.java:1374)                                at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)          at org.apache.tools.ant.Project.executeTargets(Project.java:1264)                               at org.apache.tools.ant.Main.runBuild(Main.java:818)                                            at org.apache.tools.ant.Main.startAnt(Main.java:223)                                            at org.apache.tools.ant.launch.Launcher.run(Launcher.java:284)                                  at org.apache.tools.ant.launch.Launcher.main(Launcher.java:101)                         Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.                             at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:480)                         at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)                                              at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:402)                                           ... 17 more                             Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.                   at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:128)                   at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
>         at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:475)                         ... 19 more                             Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse suppressions-xpath.xml - Premature end of file.
>         at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:292)                                 at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)
>         at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)                                at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:583)                             at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:203)                                              at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:123)                   ... 21 more                             Caused by: org.xml.sax.SAXParseException; systemId: file:/root/CheckStyleSample/suppressions-xpath.xml; lineNumber: 1; columnNumber: 1; Premature end of file.                                          at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
>         at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)                                at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)                                     at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)                                     at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1465)                                           at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:1013)
>         at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)                                at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)                                    at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)                                                      at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)                                     at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)                                   at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:81)                at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)                                 ... 26 more
> ```

There's another problem with this scenario when it generates an empty file or no suppressions given an existing suppression file. Shouldn't it display the existing suppressions on the console? Or is it expected behaviour that only newer suppressions are generated?

**Comment by rnveach** [comment]
> Run ant now and you'll find the error reproduced.

Ran again, it displayed `gen-suppressions: [java] Java Result: 1` but still displayed no error and completed successfully.

I recommend, submit a PR to our sample repo to only add an xpath match to the configuration. If it doesn't fail, then I can only assume there is some issue on your end.

> There's another problem with this scenario when it generates an empty file or no suppressions given an existing suppression file. Shouldn't it display the existing suppressions on the console? Or is it expected behaviour that only newer suppressions are generated?

This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.

**Comment by Fernal73** [comment]
> > Run ant now and you'll find the error reproduced.
> 
> Ran again, it displayed `gen-suppressions: [java] Java Result: 1` but still displayed no error and completed successfully.
> 
> I recommend, submit a PR to our sample repo to only add an xpath match to the configuration. If it doesn't fail, then I can only assume there is some issue on your end.
> 
> > There's another problem with this scenario when it generates an empty file or no suppressions given an existing suppression file. Shouldn't it display the existing suppressions on the console? Or is it expected behaviour that only newer suppressions are generated?
> 
> This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.

As far as I can see, checkstyle fails on attempting to parse existing suppressions document which requires the xml resolver class, org/xmlresolver/Resolver class which is present in the xmlresolver-5.1.1.jar which isn't downloaded by ivy as a transitive dependency for Saxon-HE-12.1.jar (which it should). If it doesn't at your end, then the bug is valid. If not, then it's an issue with ArchLinux for Termux.

**Comment by Fernal73** [comment]
> 
> This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.

I'd expect checkstyle to check if the suppressions file is empty or not before attempting to parse the document. If the suppressions weren't being generated to the console, I'd expect checkstyle to generate the suppressions document including the xml header and empty suppressions list when no suppressions are found to be generated. 
In my usecase, it makes sense to delete the existing suppressions file before generating fresh ones since I'd expect to retain the existing ones plus the new ones if found. In either case, an empty suppressions file will always crash checkstyle as it currently is.

**Comment by Fernal73** [comment]
> > This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.
> 
> I'd expect checkstyle to check if the suppressions file is empty or not before attempting to parse the document. If the suppressions weren't being generated to the console, I'd expect checkstyle to generate the suppressions document including the xml header and empty suppressions list when no suppressions are found to be generated. In my usecase, it makes sense to delete the existing suppressions file before generating fresh ones since I'd expect to retain the existing ones plus the new ones if found. In either case, an empty suppressions file will always crash checkstyle as it currently is.

I guess one could always check if no suppressions i.e., empty file is generated and delete it. Checkstyle doesn't fail if no suppressions file exists. But I think Checkstyle must incorporate a check for empty file. Or is that considered too defensive programming?

**Comment by Fernal73** [comment]
> > > This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.
> > 
> > 
> > I'd expect checkstyle to check if the suppressions file is empty or not before attempting to parse the document. If the suppressions weren't being generated to the console, I'd expect checkstyle to generate the suppressions document including the xml header and empty suppressions list when no suppressions are found to be generated. In my usecase, it makes sense to delete the existing suppressions file before generating fresh ones since I'd expect to retain the existing ones plus the new ones if found. In either case, an empty suppressions file will always crash checkstyle as it currently is.
> 
> I guess one could always check if no suppressions i.e., empty file is generated and delete it. Checkstyle doesn't fail if no suppressions file exists. But I think Checkstyle must incorporate a check for empty file. Or is that considered too defensive programming?

I guess the question is whether Checkstyle should throw a RuntimeException or whether it should exit more gracefully.

**Comment by Fernal73** [comment]
> > > Run ant now and you'll find the error reproduced.
> > 
> > 
> > Ran again, it displayed `gen-suppressions: [java] Java Result: 1` but still displayed no error and completed successfully.
> > I recommend, submit a PR to our sample repo to only add an xpath match to the configuration. If it doesn't fail, then I can only assume there is some issue on your end.
> > > There's another problem with this scenario when it generates an empty file or no suppressions given an existing suppression file. Shouldn't it display the existing suppressions on the console? Or is it expected behaviour that only newer suppressions are generated?
> > 
> > 
> > This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.
> 
> As far as I can see, checkstyle fails on attempting to parse existing suppressions document which requires the xml resolver class, org/xmlresolver/Resolver class which is present in the xmlresolver-5.1.1.jar which isn't downloaded by ivy as a transitive dependency for Saxon-HE-12.1.jar (which it should). If it doesn't at your end, then the bug is valid. If not, then it's an issue with ArchLinux for Termux.

This doesn't seem to be a bug with checkstyle per se. If you try to download Saxon-HE using a similar configuration, the xmlresolver-5.1.1.jar is missing as well. Maybe, this ought to just be documented somewhere for users who set up their classpaths using dependency managers. It appears to be either a misconfiguration with Saxon-HE or a bug in ivy while handling classified jars. It would be a good idea to test all versions of Saxon-HE since the xmlresolver-data.jar was introduced.

**Comment by romani** [comment]
Good, thanks a lot for investigation

**Comment by Fernal73** [comment]
> Good, thanks a lot for investigation

No problem, @romani. The issue couldn't be closed without raising one on Ivy. Now, it's somebody else's problem.

**Comment by linusjf** [comment]
> I emailed the Ivy user group and an issue has been raised at https://issues.apache.org/jira/plugins/servlet/mobile#issue/IVY-1642

This should be fixed by:

https://github.com/apache/ant-ivy/commit/5f11f05fe38b2fc3779d5bec6287505798bc701c
